### PR TITLE
nixos/sysctl: dynamic inotify watches or higher defaults

### PR DESCRIPTION
**Pull request by blaggacao**
Nowadays most applications require a good amount of inotify watches,
so raise our default to what other distros do. If kernel supports it
enable dynamic setting.

### CODE REVIEW

File: nixos/modules/config/sysctl.nix [hunk 1]
```diff
@@ -59,5 +59,19 @@ in
     # Disable YAMA by default to allow easy debugging.
     boot.kernel.sysctl."kernel.yama.ptrace_scope" = mkDefault 0;
 
+    # Dynamic since kernel v5.12
+    # https://github.com/torvalds/linux/commit/ac7b79fd190b02e7151bc7d2b9da692f537657f3
+    boot.kernel.sysctl."fs.inotify.max_user_instances" =
+      if lib.versionAtLeast config.boot.kernelPackages.kernel.version "5.12"
```
[the remark this answers is not in the document]
Comment (blaggacao) [reply nixos/modules/config/sysctl.nix:65]: you mean versionOlder than `5.11` ? (one arithmetic operation away from the magic number `5.12`)

File: nixos/modules/config/sysctl.nix [hunk 2]
```diff
@@ -59,5 +59,19 @@ in
...
+      then mkDefault 2147483647 # INT_MAX, dynamically limited based on available memory
+      else mkDefault 8192; # instead of 128
```
[the remark this answers is not in the document]
Comment (blaggacao) [reply nixos/modules/config/sysctl.nix:67]: > The max_user_instances sysctl limit defaults to 128 on NixOS and per user. When NixOS containers (nixos-container or declarative containers) are used, the inotify instances for root add up across containers, thus leading to exhaustion and Too many open files already for a few containers.

(https://github.com/NixOS/nixpkgs/issues/36214) &mdash; which we pretend fixing

[the remark this answers is not in the document]
Comment (nh2) [reply nixos/modules/config/sysctl.nix:67]: > (#36214) — which we pretend fixing

I don't quite understand what you mean -- but I think the anser to @alyssais's question is that the newer kernels mentioned in https://github.com/NixOS/nixpkgs/issues/36214#issuecomment-860249879 only set a dynamic kernel for `max_user_watches`, while this is `max_user_instances`, which is not dynamic.

That is, from the commit linked above:

> With inotify instances charged to memcg, the admin can simply set max_user_instances to INT_MAX and let the memcg limits of the jobs limit their inotify instances.

In other words, for `max_user_instances`, the linked commit instructs us to set the value to `INT_MAX` on newer kernels to be able to benefit from the automatic sizing (so that it isn't capped).

Does that answer your question @alyssais?

**Review by nh2 — APPROVED** [review]
Great find, change looks sensible to me.

File: nixos/modules/config/sysctl.nix (hunk 2 above)
[the remark this answers is not in the document]
Comment (blaggacao) [reply nixos/modules/config/sysctl.nix:67]: ... and in the case of older kernels, `128` might still not be enough for some use cases, so we default to `8192`, which does not come with a penalty unless those many instances are actually spawned (for reasons) on the user system.

File: nixos/modules/config/sysctl.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (blaggacao) [reply nixos/modules/config/sysctl.nix:65]: If there the is a bugfix version (independent if that actually would happen) it can break in theory. So I'm against that semantic.

[the remark this answers is not in the document]
Comment (jonringer) [reply nixos/modules/config/sysctl.nix:65]: no, means
```suggestion
      if (config.boot.kernelPackages.kernel.kernelOlder "5.12")
```

The `kernelOlder` and `kernelAtLeast` have been around since 20.09, so they should be available

[the remark this answers is not in the document]
Comment (blaggacao) [reply nixos/modules/config/sysctl.nix:65]: Thanks, it should be `kernalAtLeast` :+1: I'll use those, instead.

[the remark this answers is not in the document]
Comment (blaggacao) [reply nixos/modules/config/sysctl.nix]: Done in https://github.com/NixOS/nixpkgs/compare/cbe2715c72c66cea81815e31a642d7c852607b81..c366d5ec80c58130feff4f3e98ca9a8b03c33e1f

[the remark this answers is not in the document]
Comment (jonringer) [reply nixos/modules/config/sysctl.nix]: Yea, had the logic reversed when I wrote it :)

[the remark this answers is not in the document]
Comment (blaggacao) [reply nixos/modules/config/sysctl.nix]: It's nice, so if there should be ever a kernel specific twist to it, this will be used.

**Comment by blaggacao** [comment]
ping: this is supposed to be ready :shrug:

File: nixos/modules/config/sysctl.nix (hunk 2 above)
[the remark this answers is not in the document]
Comment (alyssais) [reply nixos/modules/config/sysctl.nix:67]: Ah, yes, I had misread the kernel commit.  I thought it would do that if no value was set, and setting a value would override it.  But we are indeed instructed to set it to INT_MAX.

File: nixos/modules/config/sysctl.nix [hunk 3]
```diff
@@ -59,5 +59,19 @@ in
...
 
+    # Dynamic since kernel v5.12
+    # https://github.com/torvalds/linux/commit/ac7b79fd190b02e7151bc7d2b9da692f537657f3
+    boot.kernel.sysctl."fs.inotify.max_user_instances" =
+      if (config.boot.kernelPackages.kernel.kernelAtLeast "5.12")
+      then mkDefault 2147483647 # INT_MAX, dynamically limited based on available memory
+      else mkDefault 8192; # instead of 128
+
+    # Dynamic since kernel v5.11
+    # https://github.com/torvalds/linux/commit/92890123749bafc317bbfacbe0a62ce08d78efb7a
```
Comment (alyssais) [inline nixos/modules/config/sysctl.nix:70]: This is a 404, so I can't check it.

**Review by alyssais — CHANGES_REQUESTED** [review]

File: nixos/modules/config/sysctl.nix (hunk 3 above)
Comment (blaggacao) [reply nixos/modules/config/sysctl.nix]: Oh, thank you there was an `a` too much at the end... :/

File: nixos/modules/config/sysctl.nix [hunk 4]
```diff
@@ -59,5 +59,19 @@ in
...
+      if (config.boot.kernelPackages.kernel.kernelAtLeast "5.12")
+      then mkDefault 2147483647 # INT_MAX, dynamically limited based on available memory
+      else mkDefault 8192; # instead of 128
+
+    # Dynamic since kernel v5.11
+    # https://github.com/torvalds/linux/commit/92890123749bafc317bbfacbe0a62ce08d78efb7
+    boot.kernel.sysctl."fs.inotify.max_user_watches" =
+      mkIf
+        (config.boot.kernelPackages.kernel.kernelAtLeast "5.11")
+        (mkDefault 1048576); # instead of 8192
```
[the remark this answers is not in the document]
Comment (blaggacao) [reply nixos/modules/config/sysctl.nix]: Oh, absolutely right: https://github.com/NixOS/nixpkgs/compare/de892e559efb0a3e93ddb8799f6d113f178bf3b4..f47bf35286bb8f968f1a717b12c7f25f603ba10b

[the remark this answers is not in the document]
Comment (blaggacao) [reply nixos/modules/config/sysctl.nix]: That didn't fix it. How can I express that with nix appropriately?

[the remark this answers is not in the document]
Comment (blaggacao) [reply nixos/modules/config/sysctl.nix]:
```nix
{
    boot.kernel.sysctl."fs.inotify.max_user_watches" = mkIf
      !(config.boot.kernelPackages.kernel.kernelAtLeast "5.11")
      (mkDefault 1048576); # instead of 8192
}
```

Is that good? That is: doesn't set `fs.inotify.max_user_watches` at all for kernel versions that are not at least 5.11?

[the remark this answers is not in the document]
Comment (alyssais) [reply nixos/modules/config/sysctl.nix]: You should use `kernelOlder` here instead of inverting `kernelAtLeast`, but the logic here looks right to me.

If you want to be extra sure you can test with `nixos-rebuild build-vm`.

[the remark this answers is not in the document]
Comment (blaggacao) [reply nixos/modules/config/sysctl.nix]: I suppose you mean a hypothetical `kernelYounger "5.11"`, but I checked that doesn't exists :disappointed:

**Comment by blaggacao** [comment]
@GrahamcOfBorg test bittorrent
@GrahamcOfBorg test boot.biosCdrom
@GrahamcOfBorg test boot.biosNetboot
@GrahamcOfBorg test boot.biosUsb
@GrahamcOfBorg test boot.uefiCdrom
@GrahamcOfBorg test boot.uefiNetboot
@GrahamcOfBorg test boot.uefiUsb
@GrahamcOfBorg test installer.bcache
@GrahamcOfBorg test installer.btrfsSimple
@GrahamcOfBorg test installer.btrfsSubvolDefault
@GrahamcOfBorg test installer.btrfsSubvols
@GrahamcOfBorg test installer.encryptedFSWithKeyfile
@GrahamcOfBorg test installer.grub1
@GrahamcOfBorg test installer.luksroot
@GrahamcOfBorg test installer.lvm
@GrahamcOfBorg test installer.separateBoot
@GrahamcOfBorg test installer.simple
@GrahamcOfBorg test installer.swraid
@GrahamcOfBorg test installer.fsroot

File: nixos/modules/config/sysctl.nix (hunk 4 above)
[the remark this answers is not in the document]
Comment (sternenseemann) [reply nixos/modules/config/sysctl.nix]:
```nix
versionAtLeast = v1: v2: !versionOlder v1 v2
```

`kernelOlder` should do exactly what you want.

[the remark this answers is not in the document]
Comment (blaggacao) [reply nixos/modules/config/sysctl.nix]: The default should apply for any `kernelVersion < 5.11`. That would be `kernelYounger`.

The suggestions about `versionOlder` / `kernelOlder`do confuse me. How are the to be parsed (in plain English)?

[the remark this answers is not in the document]
Comment (sternenseemann) [reply nixos/modules/config/sysctl.nix]: The following does what you want, right?

```nix
!(config.boot.kernelPackages.kernel.kernelAtLeast "5.11")
```

This is equivalent to

```nix
!(!(config.boot.kernelPackages.kernel.kernelAtLeast "5.11"))
```

and we can eliminate the negations.

`kernelOlder v` is not `kernelVersion <= v`, but `kernelVersion < v` which is what you seem to want to call `kernelYounger`?

[the remark this answers is not in the document]
Comment (blaggacao) [reply nixos/modules/config/sysctl.nix]: I'm getting completely lost, sry for that.

> kernelOlder v is not kernelVersion <= v, but kernelVersion < v which is what you seem to want to call kernelYounger?

Doe you mean `kernelOlder` is _actually_ `kernelYounger` and just horribly named? `kernelOlder` would be `kernelVersion > v` not `kernelVersion >= v`. And herein lies the problem:

---

- `[5.11 - <infinity>` - no defaults
- `-<infinity> - 5.11[` - defaults

If `kernelAtLeast 5.12` _is_ `[5.11 - <infinity>`, then the negation of that is everything else. That everything else is where we want to apply the defaults.

[the remark this answers is not in the document]
Comment (sternenseemann) [reply nixos/modules/config/sysctl.nix]: Seems to be doing what you want:

```
nix-repl> linuxPackages_5_12.kernelOlder "5.11"
false

nix-repl> linuxPackages_5_10.kernelOlder "5.11" 
true
```

`kernelOlder` means “Is this kernel's version older than the provided one?” and is equivalent to:

```nix
lib.versionOlder kernel.version versionToCompareTo
```

---

> If kernelAtLeast 5.11 is [5.11 - <infinity>, then the negation of that is everything else. That everything else is where we want to apply the defaults.

`kernelOlder` is precisely the negation of `kernelAtLeast` (see <https://github.com/NixOS/nixpkgs/pull/126777#discussion_r654837656>), so exactly what you want: The “older” means version lower than, so it doesn *not* include the provide version:

```
nix-repl> let kernelPackageVersion = "5.11"; in lib.versionOlder kernelPackageVersion "5.11"
false
```

[the remark this answers is not in the document]
Comment (blaggacao) [reply nixos/modules/config/sysctl.nix]: @sternenseemann Thanks for the "mental lift", it definitely helped. But this is still not correct:

```
# This is not what we want
nix-repl> linuxPackages_5_11.kernelOlder "5.11"  # there is no 5_11, but you can try with 5.10
false

# This is what we want
nix-repl> linuxPackages_5_11.kernelAtLeast "5.11"  # there is no 5_11, but you can try with 5.10
true
```

[the remark this answers is not in the document]
Comment (sternenseemann) [reply nixos/modules/config/sysctl.nix]: I thought you wanted a predicate that would exclude 5.11? If you negate kernelAtLeast then you still get the same result as with kernelOlder.

[the remark this answers is not in the document]
Comment (blaggacao) [reply nixos/modules/config/sysctl.nix]: Finally it clicked, thanks @sternenseemann !

Resolved in : https://github.com/NixOS/nixpkgs/compare/8c967c1942645b739d472f8b1a495c1aae462719..0e8af3ed9687017ba06735c9220b280793b14d06

**Comment by blaggacao** [comment]
@alyssais Can we merge this? All feedback seems to be duly attended.

**Comment by fzakaria** [comment]
This change LGTM.

File: nixos/modules/config/sysctl.nix [hunk 5]
```diff
@@ -59,5 +59,18 @@ in
     # Disable YAMA by default to allow easy debugging.
     boot.kernel.sysctl."kernel.yama.ptrace_scope" = mkDefault 0;
 
+    # Dynamic since kernel v5.12
+    # https://github.com/torvalds/linux/commit/ac7b79fd190b02e7151bc7d2b9da692f537657f3
+    boot.kernel.sysctl."fs.inotify.max_user_instances" =
+      if (config.boot.kernelPackages.kernel.kernelAtLeast "5.12")
+      then mkDefault 2147483647 # INT_MAX, dynamically limited based on available memory
```
Comment (fzakaria) [inline nixos/modules/config/sysctl.nix:66]: Can we make `boot.kernel.INT_MAX` a variable?

**Review by fzakaria — APPROVED** [review]
I think it's great.
Small nit would be making it a `let` variable or a re-usable variable for INT_MAX.
(Probably re-used elsewhere)

File: nixos/modules/config/sysctl.nix (hunk 5 above)
Comment (blaggacao) [reply nixos/modules/config/sysctl.nix:66]: I guess, it would be more like in the `stdenv`, though or `lib` or even `builtin`. I can make a quick follow-up PR once this is merged.

**Comment by maralorn** [comment]
I don‘t understand the `max_user_instances` default.

I guess it makes sense for `max_user_watches`. The commit explains that the default value changes with 5.11 so we set that same value for everything below 5.11. Seems reasonable.

For `max_user_instances` we only change the default for everything above 5.12 if I understand `kernelAtLeast` correctly. Why is that?

It might be just right, but it looks like different cases. So describing them both with "Dynamic since X" confuses me. Can you maybe elaborate that?

**Comment by blaggacao** [comment]
> Can you maybe elaborate that?

The upstream commit formulation is so intricate that I don't dare to give an approximate interpretation here inline.

https://github.com/torvalds/linux/commit/ac7b79fd190b02e7151bc7d2b9da692f537657f3

**Comment by maralorn** [comment]
Okay, do you know what harm it would do do raise that limit unconditionally?

And related: Back porting a new default is one thing, but are we sure it’s a good idea to override an existing default for current and future kernels?

**Comment by blaggacao** [comment]
Thanks for the continued discussion!

> Okay, do you know what harm it would do do raise that limit unconditionally?

I don't for sure, but I assume that it's an int value, so it cannot receive more than MAX_INT.

> And related: Back porting a new default is one thing, but are we sure it’s a good idea to override an existing default for current and future kernels?

This was inspired by how other distributions do it and was argued in the related issue: https://github.com/NixOS/nixpkgs/issues/36214

**Comment by maralorn** [comment]
> > Okay, do you know what harm it would do do raise that limit unconditionally?
> 
> I don't for sure, but I assume that it's an int value, so it cannot receive more than MAX_INT.

That’s not what I meant. I was wondering why we only do the override from 5.12 and newer.

But since then I think I understood it. I suggest adding a comment of the form:
`From version 5.12 inotify watches count against a RAM resource limit, so it’s safe to increase this limit to the maximum`
(If this is factually correct.)

> > And related: Back porting a new default is one thing, but are we sure it’s a good idea to override an existing default for current and future kernels?
> 
> This was inspired by how it was argued in the related issue: #36214, seems to be a requirement for [docker](https://github.com/NixOS/nixpkgs/pull/112472) or rather: in general name-spacing. I innocently _assume_ because of those namespaces actually count towards this maximum.
> 
> Since this is a maximum boundary and implemented as a default, it is in any case as close to a pareto-improvement as we can get (doesn't do harm to anybody, but good to many). The maximum is only hit, if you _actually_ use it.

Well the resulting harm would be if the user creates a resource exhaustion they don‘t want. The kernel has this limits for a reason. But if my phrasing above is correct, it’s fine to raise it.

File: nixos/modules/config/sysctl.nix [hunk 6]
```diff
@@ -59,5 +59,20 @@ in
     # Disable YAMA by default to allow easy debugging.
     boot.kernel.sysctl."kernel.yama.ptrace_scope" = mkDefault 0;
 
+    # Dynamic since kernel v5.12
+    # https://github.com/torvalds/linux/commit/ac7b79fd190b02e7151bc7d2b9da692f537657f3
+    # From version 5.12 inotify watches count against a RAM resource limit, so kernel
+    # mainatainers (see commit) suggest to increase this limit to the maximum
```
Comment (maralorn) [inline nixos/modules/config/sysctl.nix:65]:
```suggestion
    # maintainers (see commit) suggest to increase this limit to the maximum
```

**Comment by maralorn** [comment]
I think I am 100% in favor of the `instances` change. It does basically what the kernel devs recommend.

I am a little bit uncertain about the `watches` change. The kernel wants to establish a new, system dependent default here and this PR would raise the limit for older kernels higher then what new kernels would pick on most systems. I don‘t think this is likely to cause any problems, but maybe someone more confident can merge this.

**Comment by blaggacao** [comment]
Cleaning up my follow-up list. Please anyone feel free to pick this up.
